### PR TITLE
Add interface support to `transitiveClosure`

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Contributes to https://github.com/digital-asset/daml/issues/13324

changelog_begin
changelog_end

This work can be reused to address https://github.com/digital-asset/daml/issues/13349 and I left `TODO`s where appropriate.

While I was there, I removed a field in `Node` that was not used.

### CODE REVIEW

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/DependencyGraph.scala [hunk 1]
```diff
@@ -49,10 +55,20 @@ private[codegen] object DependencyGraph {
...
-  def transitiveClosure(decls: Map[Identifier, InterfaceType]): TransitiveClosure = {
-    val dependencies = orderedDependencies(decls)
+  def transitiveClosure(
+      serializableTypes: Map[Identifier, InterfaceType],
+      interfaces: Map[Identifier, DefInterface.FWT],
+  ): TransitiveClosure = {
+    val dependencies = orderedDependencies(serializableTypes, interfaces)
     TransitiveClosure(
-      interfaceTypes = dependencies.deps.map { case (id, Node(t, _, _)) => id -> t },
+      orderedDependencies = orderedDependencies(serializableTypes, interfaces).deps.map {
```
Comment (stefanobaghino-da) [inline language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/DependencyGraph.scala:64]: Duplicated logic that needs to be fixed.

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/TransitiveClosure.scala [hunk 2]
```diff
@@ -14,6 +14,13 @@ import com.daml.lf.iface.InterfaceType
...
-    interfaceTypes: Vector[(Identifier, InterfaceType)],
+    orderedDependencies: Vector[(Identifier, Either[DefInterface.FWT, InterfaceType])],
     errors: List[UnsupportedTypeError],
-)
+) {
+
+  lazy val (interfaces, serializableTypes) = orderedDependencies.partitionMap {
+    case (id, Left(interface)) => Left(id -> interface)
+    case (id, Right(serializable)) => Right(id -> serializable)
+  }
```
Comment (stefanobaghino-da) [inline language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/TransitiveClosure.scala:24]: This should probably be split upstream, there's little sense in keeping these two together, I think.

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/DependencyGraph.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/DependencyGraph.scala]: Addressed by https://github.com/digital-asset/daml/pull/13557/commits/9a42878f62ee9cafb6f61e7504300807a5cc9c54

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/TransitiveClosure.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/dependencygraph/TransitiveClosure.scala]: Addressed by https://github.com/digital-asset/daml/pull/13557/commits/9414b9e282b38b7c856ed4dc7ccdfeb7a7425f2f

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala [hunk 3]
```diff
@@ -248,7 +248,15 @@ object TemplateChoice {
...
     extends DefTemplate.GetChoices[Ty]
 
-object DefInterface extends FWTLike[DefInterface]
+object DefInterface extends FWTLike[DefInterface] {
+
+  implicit val `InterfaceDecl fold`: Foldable[DefInterface] =
+    new Foldable.FromFoldMap[DefInterface] {
+      override def foldMap[A, B: Monoid](fa: DefInterface[A])(f: A => B): B =
+        fa.choices.foldMap(_ foldMap f)
+    }
```
Comment (S11001001) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala:257]: 🎉

Comment (stefanobaghino-da) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala:257]: origami

**Pull request merged**